### PR TITLE
Add `ArrayElement` type

### PR DESCRIPTION
**Pull request by sindresorhus**
Closes #914 (it's stale)

### CODE REVIEW

File: test-d/array-element.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+expectType<string>(getRandomElement(['foo', 'bar', 'baz']));
+
+// Edge cases
+expectType<any>('' as ArrayElement<any>);
+expectType<never>('' as ArrayElement<never>);
+expectType<unknown>('' as ArrayElement<unknown[]>);
+
+// Non-arrays return never
+expectType<never>('' as ArrayElement<string>);
+expectType<never>('' as ArrayElement<{a: string}>);
```
Comment (som-sm) [inline test-d/array-element.ts]: Just for consistency:

```suggestion
// Basic array
expectType<string>({} as ArrayElement<string[]>);
expectType<number>({} as ArrayElement<number[]>);

// Tuple
expectType<1 | 2 | 3>({} as ArrayElement<[1, 2, 3]>);
expectType<'a' | 'b' | 'c'>({} as ArrayElement<['a', 'b', 'c']>);

// Const tuple
expectType<'foo' | 'bar' | 'baz'>({} as ArrayElement<readonly ['foo', 'bar', 'baz']>);
expectType<1 | 2 | 3>({} as ArrayElement<readonly [1, 2, 3]>);

// Mixed types
expectType<string | number>({} as ArrayElement<[string, number]>);
expectType<string | number>({} as ArrayElement<[string, number]>);

// Readonly array
expectType<string>({} as ArrayElement<readonly string[]>);
expectType<1 | 2 | 3>({} as ArrayElement<readonly [1, 2, 3]>);

// Empty array
expectType<never>({} as ArrayElement<[]>);
expectType<never>({} as ArrayElement<readonly []>);

// Union of arrays
expectType<1 | 2 | 3 | 4>({} as ArrayElement<[1, 2] | [3, 4]>);

// Function use case
declare function getRandomElement<T extends readonly unknown[]>(array: T): ArrayElement<T>;

expectType<number>(getRandomElement([1, 2, 3]));
expectType<'foo' | 'bar' | 'baz'>(getRandomElement(['foo', 'bar', 'baz'] as const));
expectType<string>(getRandomElement(['foo', 'bar', 'baz']));

// Edge cases
expectType<any>({} as ArrayElement<any>);
expectType<never>({} as ArrayElement<never>);
expectType<unknown>({} as ArrayElement<unknown[]>);

// Non-arrays return never
expectType<never>({} as ArrayElement<string>);
expectType<never>({} as ArrayElement<{a: string}>);
```

File: source/array-element.d.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
+/**
+Extracts the element type of an array or tuple.
+
+Use-cases:
+- When you need type-safe element extraction that returns `never` for non-arrays.
+- When working with arrays of unknown type where `T[number]` might not be safe.
```
Comment (som-sm) [inline source/array-element.d.ts]: > arrays of unknown type

Umm...wouldn't `T[number]` always be safe if it's an array?

File: source/array-element.d.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+- For better readability in complex type operations.
+
+Returns `never` if the type is not an array.
```
Comment (som-sm) [inline source/array-element.d.ts]: Better to add it as `Note`?

```suggestion
Note: Returns `never` if the type is not an array.
```

File: test-d/array-element.ts [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+expectType<1 | 2 | 3>(1 as ArrayElement<readonly [1, 2, 3]>);
+
+// Mixed types
+expectType<string | number>('string' as ArrayElement<[string, number]>);
+expectType<string | number>(123 as ArrayElement<[string, number]>);
+
+// Readonly array
+expectType<string>('' as ArrayElement<readonly string[]>);
+expectType<1 | 2 | 3>(1 as ArrayElement<readonly [1, 2, 3]>);
+
```
Comment (som-sm) [inline test-d/array-element.ts:23]: Add these:

```ts
// Optional and rest elements
expectType<1 | 2 | 3 | undefined>(1 as ArrayElement<[1, 2, 3?]>);
expectType<string | number>(1 as ArrayElement<[string, ...number[]]>);
expectType<1 | 2 | string>(1 as ArrayElement<[1, 2, ...string[]]>);
expectType<1 | 2 | undefined | string>(1 as ArrayElement<[1, 2?, ...string[]]>);
```

Comment (som-sm) [reply test-d/array-element.ts:23]: I guess it's fine to add `| undefined` for optional elements w/o worrying about `exactOptionalPropertyTypes` setting. This keeps the behaviour consistent with `T[number]`.

File: source/array-element.d.ts (hunk 2 above)
Comment (sindresorhus) [reply source/array-element.d.ts]: Yeah, badly phrased. Fixed.

**Comment by som-sm** [comment]
@sindresorhus Should we ask Claude to use details/summary blocks to shorten its output? It feels too much at the moment, I think.

**Comment by sindresorhus** [comment]
Already done: https://github.com/sindresorhus/type-fest/commit/cb977aad6bb9e825b5f988c56247bbe45c0f22d0
